### ISSUE TITLE
Theme prop

### ISSUE THREAD

**Issue by FongYoong** [uiwjs/react-codemirror#181]
It appears that the `theme` prop only supports `'light'|'dark'`. How do I apply [other themes](https://github.com/codemirror/CodeMirror/tree/master/theme) like the ones found in Codemirror's GitHub?

**Comment by jaywcjlove** [comment]
@FongYoong 

### `@uiw/react-codemirror@v4 +`

If you use the new version, you need to define the theme in a new way. 
`@uiw/react-codemirror@v4 +` Example: https://github.com/uiwjs/react-codemirror/blob/c276a31e6fc80c8c773b31d385786b9cef18253d/src/theme/light.ts#L1-L12

### `@uiw/react-codemirror@v3`

Used in the old version:

```jsx
import CodeMirror from '@uiw/react-codemirror';
import 'codemirror/keymap/sublime';
import 'codemirror/theme/monokai.css';

const code = 'const a = 0;';

<CodeMirror
  value={code}
  options={{
    theme: 'monokai',
    keyMap: 'sublime',
    mode: 'jsx',
  }}
/>
```

**Comment by FongYoong** [comment]
I see, thanks for giving examples on how to construct the theme. But how do we apply it to the `<CodeMirror />` component?

**Comment by FongYoong** [comment]
After examining [useCodeMirror.ts](https://github.com/uiwjs/react-codemirror/blob/c276a31e6fc80c8c773b31d385786b9cef18253d/src/useCodeMirror.ts), it seems like the themes are [concatenated](https://github.com/uiwjs/react-codemirror/blob/c276a31e6fc80c8c773b31d385786b9cef18253d/src/useCodeMirror.ts#L61) to the getExtensions array:

```javascript
// From useCodeMirror.ts, line 61
theme === 'light' ? getExtensions.push(defaultLightThemeOption) : getExtensions.push(oneDarkTheme);
```

So, if I'm not mistaken, any additional styling can be added via the [`extensions` prop](https://github.com/uiwjs/react-codemirror/blob/c276a31e6fc80c8c773b31d385786b9cef18253d/src/useCodeMirror.ts#L70):

```javascript
// From useCodeMirror.ts, line 70
getExtensions = getExtensions.concat(extensions);
```

With regards to customising the theme, [oneDarkTheme's](https://github.com/codemirror/theme-one-dark/blob/a22a40f4a66c8b63a727ef0d7f8733e84bc76d96/src/one-dark.ts#L24) source code is a good template.

For anyone who may be interested, the example below works well for me:
```javascript

// Custom theme declared in 'themes/customEditorTheme.js'
import { EditorView } from '@codemirror/view';
export const customEditorTheme = EditorView.theme({
  "&": {
    color: "#56b6c2",
    backgroundColor: "#e06c75"
  },
  ".cm-content": {
    caretColor: "#61afef"
  },
}, {dark: false});

///////////////////
// React component
import { customEditorTheme } from "../themes/customEditorTheme";

<CodeMirror
  value={"//code"}
  theme={"light"}
  extensions={[customEditorTheme]}
/>
```
***

@jaywcjlove Again, thanks for the help.

**Comment by JMLX42** [comment]
> After examining useCodeMirror.ts, it seems like the themes are concatenated to the getExtensions array:
> 
> ```
> // From useCodeMirror.ts, line 61
> theme === 'light' ? getExtensions.push(defaultLightThemeOption) : getExtensions.push(oneDarkTheme);
> ```

But that means the `dark` or `light` theme will be added anyway. And it will override the custom theme by adding some inline props.

Because of this, the following does *not* work:

```javascript
    /* Disable CodeMirror's active line highlighting. */
    "& .cm-activeLineGutter, & .cm-activeLine": {
      backgroundColor: "transparent",
    },
```

and `!important` has to be added as a workaround:

```javascript
    /* Disable CodeMirror's active line highlighting. */
    "& .cm-activeLineGutter, & .cm-activeLine": {
      backgroundColor: "transparent !important",
    },
```

It would be better to have an option to not push any theme or to set a custom theme using the `theme` prop.

**Comment by arvigeus** [comment]
@JMLX42 I added a PR for that #232  @jaywcjlove

**Comment by jaywcjlove** [comment]
@arvigeus Upgrade `v4.3.0`

**Comment by karlhorky** [comment]
Great thanks @jaywcjlove !

For anyone looking to make a custom theme with `v4.3.0+`, remember that you need to include both the `EditorView.theme()` and the `HighlightStyle.define()` in an array, if you also want to highlight the tokens in the code:

```jsx
const foreground = '#9cdcfe',
  background = '#1e1e1e',
  keyword = '#569CD6',
  controlFlowAndModuleKeywords = '#c586c0';

export const vsCodeDarkPlusTheme = EditorView.theme({
  '&': {
    color: foreground,
    backgroundColor: background,
  },
  /* ... */
});

const themeHighlightStyle = HighlightStyle.define([
  { tag: tags.keyword, color: keyword },
  {
    tag: [tags.controlKeyword, tags.moduleKeyword],
    color: controlFlowAndModuleKeywords,
  },
  /* ... */
]);

// ...
      <CodeMirror
        value={`const a = 123;`}
        theme={[themeTheme, themeHighlightStyle]}
        extensions={[javascript({ jsx: true, typescript: true })]}
      />
```

See https://github.com/codemirror/theme-one-dark/blob/main/src/one-dark.ts for a full example of this.

**Comment by karlhorky** [comment]
For the CodeMirror dependency versions `0.20.0` and above (eg. `@codemirror/view@0.20.1`), CodeMirror has changed their dependencies and imports, so you need to fix it to this:

```tsx
import { javascript } from '@codemirror/lang-javascript';
import { HighlightStyle, syntaxHighlighting } from '@codemirror/language';
import { EditorView } from '@codemirror/view';
import { css, jsx } from '@emotion/react';
import { tags } from '@lezer/highlight';

const foreground = '#9cdcfe',
  background = '#1e1e1e',
  keyword = '#569CD6',
  controlFlowAndModuleKeywords = '#c586c0';

export const theme = EditorView.theme({
  '&': {
    color: foreground,
    backgroundColor: background,
  },
  /* ... */
});

const themeHighlightStyle = HighlightStyle.define([
  { tag: tags.keyword, color: keyword },
  {
    tag: [tags.controlKeyword, tags.moduleKeyword],
    color: controlFlowAndModuleKeywords,
  },
  /* ... */
]);

// ...

<CodeMirror
  value={`const a = 123;`}
  theme={[theme, syntaxHighlighting(themeHighlightStyle)]}
  extensions={[javascript({ jsx: true, typescript: true })]}
/>
```

**Comment by karlhorky** [comment]
@jaywcjlove I would recommend that you upgrade to `0.20.0+` of all CodeMirror packages - there are a few changes in the structure, exports and interdependencies, and some packages are now obsolete.

cc @marijnh

**Comment by karlhorky** [comment]
Nice, looks like it's already done in #301 (and published in `4.7.0`) 🙌🎉

Fast, thanks @jaywcjlove !
